src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,895 @@
+            if let Some(s) = stage.states.get(&split_id) {
+                target_state.insert(split_id, s.clone());
+            } else {
+                split_changed = true;
+
+                let backfill_state = self
+                    .backfill_state_store
+                    .try_recover_from_state_store(&split_id)
+                    .await?;
+                match backfill_state {
+                    None => {
+                        // Newly added split. We don't need to backfill.
+                        target_state.insert(split_id, BackfillState::Finished);
+                    }
+                    Some(backfill_state) => {
+                        // Migrated split. Backfill if unfinished.
+                        match backfill_state {
+                            BackfillState::Backfilling(ref offset) => {
+                                if let Some(offset) = offset {
+                                    split.update_in_place(offset.clone())?;
+                                }
+                                stage.unfinished_splits.push(split);
+                            }
+                            BackfillState::SourceCachingUp(_) | BackfillState::Finished => {}
+                        }
+                        target_state.insert(split_id, backfill_state);
+                    }
+                }
+            }
+        }
+
+        // Checks dropped splits
+        for existing_split_id in stage.states.keys() {
+            if !target_state.contains_key(existing_split_id) {
+                tracing::info!("split dropping detected: {}", existing_split_id);
+                split_changed = true;
+            }
+        }
+
+        if split_changed {
+            tracing::info!(
+                actor_id = self.actor_ctx.id,
+                state = ?target_state,
+                "apply split change"
+            );
+
+            stage
+                .unfinished_splits
+                .retain(|split| target_state.get(split.id().as_ref()).is_some());
+
+            let dropped_splits = stage
+                .states
+                .extract_if(|split_id, _| target_state.get(split_id).is_none())
+                .map(|(split_id, _)| split_id);
+
+            if should_trim_state {
+                // trim dropped splits' state
+                self.backfill_state_store.trim_state(dropped_splits).await?;
+            }
+
+            stage.states = target_state;
+        }
+
+        Ok(split_changed)
+    }
+
+    /// For split change during forward stage, all newly added splits should be already finished.
+    // We just need to update the state store if necessary.
+    async fn apply_split_change_forward_stage(
+        &mut self,
+        split_assignment: &HashMap<ActorId, Vec<SplitImpl>>,
+        splits: &mut HashSet<SplitId>,
+        should_trim_state: bool,
+    ) -> StreamExecutorResult<()> {
+        self.metrics
+            .source_split_change_count
+            .with_label_values(
```

Comment:
Recommend to call `with_label_values` once and keep the metric in `self`, which seems to be the right usage of it.